src/persist-client/src/critical.rs

```diff
@@ -322,4 +341,41 @@ mod tests {
         let container: Container = serde_json::from_value(json).expect("deserializable");
         assert_eq!(container.reader_id, id);
     }
+
+    // Verifies that the handle update its view of the opaque token correctly
+    #[tokio::test]
+    async fn opaque_token_view() {
+        let client = crate::tests::new_test_client().await;
+
+        let shard_id = crate::ShardId::new();
+
+        let mut since = client
+            .open_critical_since::<(), (), u64, i64, i64>(
+                shard_id,
+                crate::PersistClient::CONTROLLER_CRITICAL_SINCE,
+            )
+            .await
+            .expect("codec mismatch");
+
+        // The token must be initialized to the default value
+        assert_eq!(since.opaque(), &0);
+
+        // Check that the first operation never gets rate limited
+        since
+            .compare_and_downgrade_since(&0, (&5, &Antichain::from_elem(0)))
+            .await
+            .unwrap();
+
+        // Now let's open another handle with the same reader id and shard id
+        let since2 = client
+            .open_critical_since::<(), (), u64, i64, i64>(
+                shard_id,
+                crate::PersistClient::CONTROLLER_CRITICAL_SINCE,
+            )
+            .await
+            .expect("codec mismatch");
+
+        // Our view of the token must be updated now
+        assert_eq!(since2.opaque(), &5);
```

Comment:
@danhhz @pH14 I checked in this test which I came up while trying to debug the CI failures and this assert fires. Is this expected?

Reply:
hm, right now we only update the `since` and `opaque` if `PartialOrder::less_than(state_since, new_since)` but I wonder if that ought to be `less_equal` instead

Reply:
yeah I think that should be fine. I've got a few small tweaks incl this one I can PR

Reply:
https://github.com/MaterializeInc/materialize/pull/15989